### PR TITLE
DEP: Deprecate size-one ragged array coercion

### LINKED ISSUE #16939: ValueError: invalid __array_struct__ when assigning to object-dtype ndarray
Hi. Pandas is broken with numpy master :)

### Reproducing code example:

```python

In [1]: import numpy as np
In [2]: out = np.empty(3, dtype="object")
In [3]: out[:] = [np.int64, np.int64, np.int64]
```
### Error message:

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-d20bd885ed9b> in <module>
----> 1 out[:] = [np.int64, np.int64, np.int64]

ValueError: invalid __array_struct__
```

### Numpy/Python version information:

Installed from a wheel with `pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy`

```
1.20.0.dev0+2d12d0c 3.7.6 (default, Dec 30 2019, 19:38:28)
[Clang 11.0.0 (clang-1100.0.33.16)]
```

I don't immediately see what commit could have caused this. https://github.com/numpy/numpy/pull/16850 mentions object-dtype scalars, but doesn't look quite right.

### PR DESCRIPTION
**Pull request by seberg**
Previously, code such as:
```
np.array([0, np.array([0])], dtype=np.int64)
```
worked by discovering the array as having a shape of `(2,)` and
then using `int(np.array([0]))` (which currently succeeds).
This is also a ragged array, and thus deprecated here earlier on.
(As a detail, in the new code the assignment should occur as
an array assignment and not using the `__int__`/`__float__`
Python protocols.)

Two details to note:

1. This still skips the deprecation for sequences which are not
   array-likes.  We assume that sequences will always fail the
   conversion to a number.
2. The conversion to a number (or string, etc.) may still fail
   after the DeprecationWarning is given.  This is not ideal, but
   narrowing it down seems tedious, since the actual assignment
   happens in a later step.
   I.e. `np.array([0, np.array([None])], dtype=np.int64)` will give
   the warning, but then fail anyway, since the array cannot be
   assigned.

Closes gh-16939

### CODE REVIEW

**Comment by seberg** [comment]
I am honestly a bit surprised this missed the old ragged-array deprecation, but I don't recall exactly, maybe the old ragged array deprecation was wired into the dtype discovery chunk and not activated when a dtype is passed in.

**Comment by eric-wieser** [comment]
Ping @nschloe, since this is a much weaker form of something you wanted to deprecate.

File: numpy/core/src/multiarray/array_coercion.c [hunk 1]
```diff
@@ -1221,7 +1221,58 @@ PyArray_DiscoverDTypeAndShape(
...
+                        deprecate_single_element_ragged = 0;
+                        break;
+                    }
+                    deprecate_single_element_ragged = 1;
+                }
+                current = current->next;
+            }
+
+            if (deprecate_single_element_ragged) {
+                /* Deprecated 2019-07-24, NumPy 1.20 */
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/array_coercion.c:1264]: Are you sure this is the right date? ;)

Comment (seberg) [inline numpy/core/src/multiarray/array_coercion.c:1264]:
```suggestion
                /* Deprecated 2020-07-24, NumPy 1.20 */
```

**Comment by charris** [comment]
Let's give it a shot. Thanks Sebastian.

**Pull request merged**